planner/core/common_plans.go

```diff
@@ -297,36 +287,67 @@ func (e *Execute) OptimizePreparedPlan(ctx context.Context, sctx sessionctx.Cont
 		}
 		prepared.SchemaVersion = is.SchemaMetaVersion()
 	}
-	err := e.getPhysicalPlan(ctx, sctx, is, preparedObj)
+	err = e.getPhysicalPlan(ctx, sctx, is, preparedObj)
 	if err != nil {
 		return err
 	}
 	e.SnapshotTS = snapshotTS
+	e.TxnScope = txnScope
+	e.IsStaleness = isStaleness
 	e.Stmt = prepared.Stmt
-	e.handleStalenessOption(sctx)
 	return nil
 }
 
-func (e *Execute) handleStalenessOption(sctx sessionctx.Context) {
+func (e *Execute) handleExecuteBuilderOption(sctx sessionctx.Context,
+	preparedObj *CachedPrepareStmt) (snapshotTS uint64, txnScope string, isStaleness bool, err error) {
+	snapshotTS = 0
+	txnScope = oracle.GlobalTxnScope
+	isStaleness = false
+	err = nil
 	vars := sctx.GetSessionVars()
 	readTS := vars.TxnReadTS.PeakTxnReadTS()
-	// If readTS > 0, it means we are handling the following case:
-	// 1. set transaction read only as of timestamp ts
-	// 2. execute prepared statement
 	if readTS > 0 {
-		e.IsStaleness = true
-		e.SnapshotTS = vars.TxnReadTS.UseTxnReadTS()
-		e.TxnScope = config.GetTxnScopeFromConfig()
+		// It means we meet following case:
+		// 1. set transaction read only as of timestamp ts
+		// 2. execute prepared statement with different as of ts
+		if preparedObj.SnapshotTSEvaluator != nil {
+			err = ErrAsOf.FastGenWithCause("as of timestamp can't be set after set transaction read only as of.")
+			return
+		}
+		// If readTS > 0, it means we are handling the following case:
+		// 1. set transaction read only as of timestamp ts
+		// 2. execute prepared statement
+		snapshotTS = vars.TxnReadTS.UseTxnReadTS()
+		isStaleness = true
+		txnScope = config.GetTxnScopeFromConfig()
+		return
+	}
+	if preparedObj.SnapshotTSEvaluator != nil {
+		if vars.InTxn() {
+			err = ErrAsOf.FastGenWithCause("as of timestamp can't be set in transaction.")
```

Comment:
Ditto. Maybe talk about it is a prepared stale SQL, executed in the transaction.